File: spring-boot-commons-util/src/test/java/com/github/springframework/boot/commons/util/DateTimeTest.java
package com.github.springframework.boot.commons.util;

import com.github.springframework.boot.commons.util.time.DateTime;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.TimeZone;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class DateTimeTest {

	@BeforeAll
	static void setup() {
		TimeZone.setDefault(TimeZone.getTimeZone("Asia/Shanghai"));
	}

	@Test
	void testNewInstance() {
		UnsupportedOperationException e = assertThrows(UnsupportedOperationException.class, DateTime::new);
		assertEquals("Utility class should not be instantiated", e.getMessage());
	}

	@ParameterizedTest
	@CsvSource({
		"2022-11-07 00:00:00, Asia/Phnom_Penh, 2022-11-06 23:00:00",
		"2022-11-07 00:00:00, America/New_York, 2022-11-06 11:00:00"
	})
	void testToLocalDateTime(String datetimeWithDefaultZone, String targetZoneId, String expected) throws ParseException {
		ZoneId zoneId = ZoneId.of(targetZoneId);
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DateTime.DEFAULT_DATE_TIME_PATTERN);

		LocalDateTime result = DateTime.toLocalDateTime(datetimeWithDefaultZone, formatter, zoneId);
		assertEquals(expected, result.format(formatter));

		SimpleDateFormat dateFormat = new SimpleDateFormat(DateTime.DEFAULT_DATE_TIME_PATTERN);
		Date parsedDate = dateFormat.parse(datetimeWithDefaultZone);
		result = DateTime.toLocalDateTime(parsedDate, zoneId);
		assertEquals(expected, result.format(formatter));

		LocalDateTime localDateTime = LocalDateTime.parse(datetimeWithDefaultZone, formatter);
		result = DateTime.toLocalDateTime(localDateTime, zoneId);
		assertEquals(expected, result.format(formatter));

		LocalDate localDate = localDateTime.toLocalDate();
		LocalDate zonedLocalDate = DateTime.toLocalDate(localDate, zoneId);
		assertEquals(expected.substring(0, 10), zonedLocalDate.toString());

		Date date = DateTime.toDate(localDateTime);
		result = DateTime.toLocalDateTime(date, zoneId);
		assertEquals(expected, result.format(formatter));
	}

	@ParameterizedTest
	@CsvSource({"2022-11-07, Asia/Phnom_Penh, 2022-11-06", "2022-11-07, America/New_York, 2022-11-06"})
	void testToLocalDate(String datetimeWithDefaultZone, String targetZoneId, String expected) {
		ZoneId zoneId = ZoneId.of(targetZoneId);
		DateTimeFormatter formatter = DateTimeFormatter.ISO_LOCAL_DATE;
		LocalDate localDate = DateTime.toLocalDate(datetimeWithDefaultZone, formatter, zoneId);
		assertEquals(expected, localDate.format(formatter));
	}

	@ParameterizedTest
	@CsvSource({"2022-11-07 00:00:00, true", "This is not a LocalDateTime string, false"})
	void testIsParseableLocalDateTime(String datetime, boolean expected) {
		assertEquals(expected, DateTime.isParseableLocalDateTime(datetime));
	}

	@ParameterizedTest
	@CsvSource({"2022-11-07, true", "This is not a LocalDate string, false"})
	void testIsParseableLocalDate(String date, boolean expected) {
		assertEquals(expected, DateTime.isParseableLocalDate(date));
	}

	@ParameterizedTest
	@CsvSource({
		"UTC+7, Asia/Jakarta, true",
		"UTC+8, Asia/Jakarta, false",
		"UTC+8, Asia/Makassar, true",
		"UTC+9, Asia/Makassar, false",
		"UTC+8, Asia/Ujung_Pandang, true",
		"UTC+9, Asia/Ujung_Pandang, false"
	})
	void testIsSameZone(String zoneId, String anotherZoneId, boolean expected) {
		assertEquals(expected, DateTime.isSameZone(ZoneId.of(zoneId), ZoneId.of(anotherZoneId)));
	}

	@Test
	void testMinAndMaxLocalDateTime() {
		DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DateTime.DEFAULT_DATE_TIME_PATTERN);
		LocalDateTime datetime1 = LocalDateTime.parse("2024-11-07 00:00:00", formatter);
		LocalDateTime datetime2 = LocalDateTime.parse("2024-11-06 23:00:00", formatter);
		LocalDateTime datetime3 = LocalDateTime.parse("2023-11-07 00:00:00", formatter);
		LocalDateTime datetime4 = LocalDateTime.parse("2023-11-06 23:00:00", formatter);
		LocalDateTime datetime5 = LocalDateTime.parse("2024-11-11 00:00:00", formatter);

		LocalDateTime max = DateTime.max(datetime1, datetime2, datetime3, datetime4, datetime5);
		assertEquals(max, datetime5);

		LocalDateTime min = DateTime.min(datetime1, datetime2, datetime3, datetime4, datetime5);
		assertEquals(min, datetime4);
	}

	@Test
	void testMinAndMaxLocalDate() {
		LocalDate date1 = LocalDate.parse("2024-11-07");
		LocalDate date2 = LocalDate.parse("2024-11-06");
		LocalDate date3 = LocalDate.parse("2023-11-07");
		LocalDate date4 = LocalDate.parse("2023-11-06");
		LocalDate date5 = LocalDate.parse("2024-11-11");

		LocalDate max = DateTime.max(date1, date2, date3, date4, date5);
		assertEquals(max, date5);

		LocalDate min = DateTime.min(date1, date2, date3, date4, date5);
		assertEquals(min, date4);
	}

}

File: spring-boot-commons-base/src/test/java/com/github/awesome/springboot/commons/base/DateTimeTest.kt
package com.github.awesome.springboot.commons.base

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class DateTimeTest {

    @BeforeEach
    fun setup() {
        TimeZone.setDefault(TimeZone.getTimeZone("Asia/Shanghai"))
    }

    @Test
    fun testNewInstance() {
        val e = assertThrows<UnsupportedOperationException> {
            DateTime()
        }
        assertEquals("Utility class should not be instantiated", e.message)
    }

    @Test
    fun testAtZone() {
        // Test null parameter
        val e1 = assertThrows<IllegalArgumentException> {
            DateTime.atZone(null, ZoneId.systemDefault())
        }
        assertEquals("datetime or targetZoneId is null", e1.message)

        val e2 = assertThrows<IllegalArgumentException> {
            DateTime.atZone(LocalDateTime.now(), null)
        }
        assertEquals("datetime or targetZoneId is null", e2.message)

        // Prepare data
        val zonedDateTime = LocalDateTime.of(2024, 11, 7, 0, 0, 0).atZone(ZoneId.systemDefault())
        val targetZoneId = ZoneId.of("Asia/Jakarta")

        // Date
        val date = Date.from(zonedDateTime.toInstant())
        val result1 = DateTime.atZone(date, targetZoneId)
        val simpleDateFormat = SimpleDateFormat(DateTime.DEFAULT_DATE_TIME_PATTERN)
        assertEquals("2024-11-06 23:00:00", simpleDateFormat.format(result1))

        // LocalDate
        val localDate = zonedDateTime.toLocalDate()
        val result2 = DateTime.atZone(localDate, targetZoneId)
        assertEquals("2024-11-06", result2.toString())

        // LocalDateTime
        val localDateTime = zonedDateTime.toLocalDateTime()
        val result3 = DateTime.atZone(localDateTime, targetZoneId).toString()
        assertEquals("2024", result3.substring(0, 4))
        assertEquals("11", result3.substring(5, 7))
        assertEquals("06", result3.substring(8, 10))
        assertEquals("23", result3.substring(11, 13))

        // LocalDate as string
        val result4 = DateTime.atZone("2024-11-07", targetZoneId)
        assertEquals("2024-11-06", result4.toString())

        // LocalDateTime as string
        val result5 = DateTime.atZone("2024-11-07 00:00:00.000", targetZoneId).toString()
        assertEquals("2024", result5.substring(0, 4))
        assertEquals("11", result5.substring(5, 7))
        assertEquals("06", result5.substring(8, 10))
        assertEquals("23", result5.substring(11, 13))

        // Unsupported datetime format
        val e3 = assertThrows<IllegalArgumentException> {
            DateTime.atZone("notParsableDateTimeString", targetZoneId)
        }
        assertEquals("Unsupported date time format: notParsableDateTimeString", e3.message)

        // Unsupported type
        val e4 = assertThrows<IllegalArgumentException> {
            DateTime.atZone(System.currentTimeMillis(), targetZoneId)
        }
        assertEquals("Unsupported type: java.lang.Long", e4.message)
    }

    @Test
    fun testIsSameZone() {
        val e1 = assertThrows<IllegalArgumentException> {
            DateTime.isSameZone(null, ZoneId.systemDefault())
        }
        assertEquals("zoneId or anotherZoneId is null", e1.message)

        val e2 = assertThrows<IllegalArgumentException> {
            DateTime.isSameZone(ZoneId.systemDefault(), null)
        }
        assertEquals("zoneId or anotherZoneId is null", e2.message)

        val defaultZoneId = ZoneId.systemDefault();
        assertTrue { DateTime.isSameZone(defaultZoneId, defaultZoneId) }
        assertTrue { DateTime.isSameZone(ZoneId.systemDefault(), ZoneId.systemDefault()) }
        assertTrue { DateTime.isSameZone(ZoneId.of("Asia/Jakarta"), ZoneId.of("Asia/Jakarta")) }
        assertFalse { DateTime.isSameZone(ZoneId.of("Asia/Jakarta"), ZoneId.of("Asia/Shanghai")) }
    }

    @Test
    fun testMaxLocalDateTime() {
        val localDateTime1 = LocalDateTime.of(2024, 11, 7, 0, 0, 0)
        val localDateTime2 = LocalDateTime.of(2024, 11, 7, 0, 0, 1)
        val localDateTime3 = LocalDateTime.of(2024, 11, 6, 23, 59, 59)
        assertEquals(localDateTime2, DateTime.max(localDateTime1, localDateTime2, localDateTime3))
    }

    @Test
    fun testMinLocalDateTime() {
        val localDateTime1 = LocalDateTime.of(2024, 11, 7, 0, 0, 0)
        val localDateTime2 = LocalDateTime.of(2024, 11, 7, 0, 0, 1)
        val localDateTime3 = LocalDateTime.of(2024, 11, 6, 23, 59, 59)
        assertEquals(localDateTime3, DateTime.min(localDateTime1, localDateTime2, localDateTime3))
    }

    @Test
    fun testMaxLocalDate() {
        val localDate1 = LocalDate.of(2024, 11, 7)
        val localDate2 = LocalDate.of(2024, 11, 6)
        val localDate3 = LocalDate.of(2024, 11, 8)
        assertEquals(localDate3, DateTime.max(localDate1, localDate2, localDate3))
    }

    @Test
    fun testMinLocalDate() {
        val localDate1 = LocalDate.of(2024, 11, 8)
        val localDate2 = LocalDate.of(2024, 11, 7)
        val localDate3 = LocalDate.of(2024, 11, 6)
        assertEquals(localDate3, DateTime.min(localDate1, localDate2, localDate3))
    }

    @Test
    fun testIsWithinDays() {
        val e1 = assertThrows<IllegalArgumentException> {
            DateTime.isWithinDays("notParsableDateTimeString", "2024-11-07 00:00:00", 1)
        }
        assertEquals("For input 'startDateTime': notParsableDateTimeString", e1.message)

        val e2 = assertThrows<IllegalArgumentException> {
            DateTime.isWithinDays("2024-11-07 00:00:00", "notParsableDateTimeString", 1)
        }
        assertEquals("For input 'endDateTime': notParsableDateTimeString", e2.message)

        val e3 = assertThrows<IllegalArgumentException> {
            DateTime.isWithinDays("2024-11-07 00:00:00", "2024-11-07 00:00:00", -1)
        }
        assertEquals("For input 'days': -1", e3.message)

        val e4 = assertThrows<IllegalArgumentException> {
            DateTime.isWithinDays("2024-11-07 00:00:00", "2024-11-07 00:00:00", 0)
        }
        assertEquals("For input 'days': 0", e4.message)

        val e5 = assertThrows<IllegalArgumentException> {
            DateTime.isWithinDays("2024-11-07 00:00:01", "2024-11-07 00:00:00", 1)
        }
        assertEquals("For input 'startDateTime': 2024-11-07 00:00:01 should be before 'endDateTime': 2024-11-07 00:00:00", e5.message)

        assertTrue { DateTime.isWithinDays("2024-11-07 00:00:00", "2024-11-07 00:00:00", 1) }
        assertTrue { DateTime.isWithinDays("2024-11-07 00:00:00", "2024-11-07 23:59:59", 1) }
        assertTrue { DateTime.isWithinDays("2024-11-07 00:00:00", "2024-11-08 00:00:00", 1) }
        assertTrue { DateTime.isWithinDays("2024-11-07 10:24:00", "2024-11-07 23:59:59", 1) }
        assertTrue { DateTime.isWithinDays("2024-11-07 10:24:00", "2024-11-08 10:24:00", 1) }
        assertFalse { DateTime.isWithinDays("2024-11-07 10:24:48", "2024-11-08 10:24:59", 1) }
        assertFalse { DateTime.isWithinDays("2024-11-07 10:24:48", "2024-11-11 10:24:48", 2) }
    }

    @Test
    fun testStartOfToday() {
        val startOfToday = DateTime.startOfToday()
        val now = LocalDateTime.now()
        assertEquals(now.year, startOfToday.substring(0, 4).toInt())
        assertEquals(now.monthValue, startOfToday.substring(5, 7).toInt())
        assertEquals(now.dayOfMonth, startOfToday.substring(8, 10).toInt())
        assertEquals(0, startOfToday.substring(11, 13).toInt())
        assertEquals(0, startOfToday.substring(14, 16).toInt())
        assertEquals(0, startOfToday.substring(17, 19).toInt())
    }

    @Test
    fun testEndOfToday() {
        val endOfToday = DateTime.endOfToday()
        val now = LocalDateTime.now()
        assertEquals(now.year, endOfToday.substring(0, 4).toInt())
        assertEquals(now.monthValue, endOfToday.substring(5, 7).toInt())
        assertEquals(now.dayOfMonth, endOfToday.substring(8, 10).toInt())
        assertEquals(23, endOfToday.substring(11, 13).toInt())
        assertEquals(59, endOfToday.substring(14, 16).toInt())
        assertEquals(59, endOfToday.substring(17, 19).toInt())
    }

}
